src/ServiceControl.Monitoring/QueueLength/QueueLengthCalculator.cs

```diff
@@ -3,45 +3,81 @@
+
+            return receivedSequences.Keys.Where(k => k.SessionKey == key).ToList();
         }
 
-        public Dictionary<string, long> GetQueueLengths()
+        public Dictionary<VirtualQueueId, double> GetQueueLengths()
         {
-            var queueLengths = new Dictionary<string, long>();
+            var queueLengths = new Dictionary<VirtualQueueId, double>();
 
-            foreach (var kvp in sessionKeyToQueueMapping)
+            foreach (var virtualQueueId in receivedSequences.Keys)
             {
-                var queueName = kvp.Value;
-                long queueLength;
-                queueLengths.TryGetValue(queueName, out queueLength);
-
-                long sent;
-                sentSequences.TryGetValue(kvp.Key, out sent);
-                var received = receivedSequences[kvp.Key];
+                double sent;
+                sentSequences.TryGetValue(virtualQueueId.SessionKey, out sent);
 
-                queueLength += sent;
-                queueLength -= received;
+                var received = receivedSequences[virtualQueueId];
 
-                queueLength = Math.Max(0, queueLength);
-                queueLengths[queueName] = queueLength;
+                //TODO: make sure that if we get two reports for the same virtual queue we keep the max values only
+                queueLengths.Add(virtualQueueId, Math.Max(0, sent - received));
             }
+
             return queueLengths;
         }
 
-        ConcurrentDictionary<string, long> receivedSequences = new ConcurrentDictionary<string, long>();
-        ConcurrentDictionary<string, long> sentSequences = new ConcurrentDictionary<string, long>();
-        ConcurrentDictionary<string, string> sessionKeyToQueueMapping = new ConcurrentDictionary<string, string>();
+        ConcurrentDictionary<VirtualQueueId, double> receivedSequences = new ConcurrentDictionary<VirtualQueueId, double>();
+        ConcurrentDictionary<string, double> sentSequences = new ConcurrentDictionary<string, double>();
+    }
+
+    public class VirtualQueueId
+    {
+        public string QueueName { get; set; }
+        public string SessionKey { get; set; }
+        public string EndpointName { get; set; }
+
+        protected bool Equals(VirtualQueueId other)
+        {
+            return string.Equals(QueueName, other.QueueName) && string.Equals(SessionKey, other.SessionKey) && string.Equals(EndpointName, other.EndpointName);
+        }
+
+        public override bool Equals(object obj)
+        {
+            if (ReferenceEquals(null, obj)) return false;
+            if (ReferenceEquals(this, obj)) return true;
+            if (obj.GetType() != this.GetType()) return false;
+            return Equals((VirtualQueueId) obj);
+        }
+
+        public override int GetHashCode()
+        {
+            unchecked
+            {
+                var hashCode = (QueueName != null ? QueueName.GetHashCode() : 0);
```

Comment:
Just thinking, isn't string.Length sufficient enough for this hash? Have you considered other implementations?